### PR TITLE
feat: add bibliography support

### LINKED ISSUE #147: Implement references / bibliography
In mathlib (3) we had a syntax for citing references stored in a `.bib` file and displaying them in a custom page https://leanprover-community.github.io/mathlib_docs/references.html.

We previously used the `pybtex` python library for obtaining the data from the bib file, this could be replaced by a call to a command line tool to convert to a more easily lean-parseable format (e.g. using `pybtex`)

### PR DESCRIPTION
**Pull request by acmepjz**
This PR adds bibliography support to doc-gen4. It includes:

- `DocGen4/Output/References.lean` which contains generic function support for bibliography. It is independent of the actual implementation of bib file formatting.
- `DocGen4/Output/Bibtex.lean` which contains an implementation of bib file formatting by calling external Lean library [BibtexQuery](https://github.com/dupuisf/BibtexQuery).
- Changes in `DocGen4/Output/DocString.lean` which analyzes citations in markdown text, inserting and modifying citation links, and collect back references.
- Changes in various output files to add back references collecting feature.
- Add a subcommand `bibPrepass` and change the build file accordingly.

This PR has all commit history squashed. To see more history, see the previous obsolete PR #200 which also includes an optional bibtex parser by calling external program `pybtex`.

Closes #147.

### CODE REVIEW

File: DocGen4/Output/Module.lean [hunk 1]
```diff
@@ -199,6 +199,9 @@ def moduleToHtml (module : Process.Module) : HtmlM Html := withTheReader SiteBas
   let relevantMembers := module.members.filter Process.ModuleMember.shouldRender
   let memberDocs ← relevantMembers.mapM (moduleMemberToHtml module.name)
   let memberNames := filterDocInfo relevantMembers |>.map DocInfo.getName
+  letI : MonadLift BaseHtmlM HtmlM := {
```
Comment (hargoniX) [inline DocGen4/Output/Module.lean:202]: Is this needed? I would have expected Lean to infer this instance on its own.

**Review by hargoniX — COMMENTED** [review]
Okay this looks good to me now apart from the one nitpick. I'll verify myself that this works on mathlib and then merge this.

Note that because this does not collect references.bib of subprojects this won't work in the mathlib4_docs CI without changes, I'll add a temporary hack to make this work out but I hope that we can add collecting subproject references.bib in the future?

**Review by hargoniX — CHANGES_REQUESTED** [review]
Running this on mathlib almost instantly throws:
```
info: stderr:
uncaught exception: failed to run texDiacritics on '\url{https://leanprover.github.io/theorem_proving_in_lean/}' at pos 4: unsupported command: '\url'
error: external command './././../doc-gen4/.lake/build/bin/doc-gen4' exited with code 1
```
Please verify that your branch actually works on mathlib.

**Comment by acmepjz** [comment]
> Please verify that your branch actually works on mathlib.

Ah, sorry for that. I forgot I have slightly local modifications for the test file. Specifically, `\url` and `\_` are the commands in that which are not supported by current parser.

- For the `\url` support an upgrade to the parser is necessary; I'll work on it later. Meanwhile I think it's better to use the standard `url` field of the bib file.
- For the `\_` it can be added quickly to BibtexQuery. But In my opinion it's inappropriate to use this command in the `url` and `doi` field. In some other entries, there are also `_` but which are not escaped.

I'll open a PR to mathlib. What's your opinion?

File: DocGen4/Output/Module.lean (hunk 1 above)
Comment (acmepjz) [reply DocGen4/Output/Module.lean:202]: I checked again and unfortunately it's still needed, as we already discussed https://github.com/leanprover/doc-gen4/pull/200#discussion_r1650098285

**Comment by acmepjz** [comment]
> I hope that we can add collecting subproject references.bib in the future?

I think it's possible, but I think this needs to be done mostly in build script. The build script collects all paths of bib files, feed them to the doc-gen4. I'm not familiar with that part.

**Comment by acmepjz** [comment]
The PR on mathlib was merged, so the current code should work on mathlib again. Meanwhile I'll submit a PR to BibtexQuery to refactor the TeX command processing. This will take some days. Do you think it's good to merge the current code to doc-gen4?

File: DocGen4/Output/Module.lean (hunk 1 above)
Comment (hargoniX) [reply DocGen4/Output/Module.lean:202]: Okay I know how to fix it, will merge this then fix it myself.

**Pull request merged**